apps/api/src/domain/models/bibliographic-citation/book-bibliographic-citation/entities/book-bibliographic-citation.from-event-history.spec.ts

```diff
@@ -0,0 +1,252 @@
+                });
+            });
+        });
+
+        describe(`when the book bibliographic citation has been published`, () => {
+            const result = BookBibliographicCitation.fromEventHistory(
+                bookPublished.as(aggregateCompositeIdentifier),
+                bookBibliographicCitationId
+            );
+
+            expect(result).toBeInstanceOf(BookBibliographicCitation);
+
+            const bookBibliographicCitation = result as BookBibliographicCitation;
+
+            expect(bookBibliographicCitation.published).toBe(true);
+        });
+
+        describe(`when a user has been granted read access`, () => {
+            it('should now allow access to the user', () => {
+                const result = BookBibliographicCitation.fromEventHistory(
+                    readAccessGranted.as(aggregateCompositeIdentifier),
+                    bookBibliographicCitationId
+                );
+
+                expect(result).toBeInstanceOf(BookBibliographicCitation);
+
+                const book = result as BookBibliographicCitation;
+
+                expect(book.queryAccessControlList.canUser(userId)).toBe(true);
+            });
+        });
+    });
+
+    describe(`when the event history is invalid`, () => {
+        describe(`when there are no events for the given Book Bibliographic Citation`, () => {
+            it(`should return Not Found`, () => {
+                const differentId = buildDummyUuid(987);
+
+                const result = BookBibliographicCitation.fromEventHistory(
+                    readAccessGranted.as({
+                        type: AggregateType.bibliographicCitation,
+                        id: differentId,
+                    }),
+                    bookBibliographicCitationId
+                );
+
+                expect(result).toBe(NotFound);
+            });
+        });
+
+        describe(`when the creation event is missing`, () => {
+            it(`should throw`, () => {
+                const eventSource = () => {
+                    BookBibliographicCitation.fromEventHistory(
+                        // omit the creation event
+                        bookPublished.as(aggregateCompositeIdentifier).slice(1),
+                        bookBibliographicCitationId
+                    );
+                };
+
+                expect(eventSource).toThrow();
+            });
+        });
+
+        // This could happen if we change our invariant rules or botch event versioning
+        describe(`when there is a creation event with invalid data`, () => {
+            it(`should return an invariant validation error`, () => {
+                const invalidEventHistory = new TestEventStream()
+                    .andThen<BookBibliographicCitationCreated>({
+                        type: 'BOOK_BIBLIOGRAPHIC_CITATION_CREATED',
+                        payload: {
+                            // this is not allowed
+                            title: 99 as unknown as string,
```

Comment:
create named var for this case?

Reply:
I prefer not to given the need for the cast. It should be readily apparent what the purpose of this is. Note that 99 is not really a magic number. It's a random non-string value that's intended to invalidate invariants. I will update the comment to be more explicit about this point, though.

Reply:
Done.